### PR TITLE
Build docker service in code, to cope with variations

### PR DESCRIPTION
**Pull request by justinsb**
The docker systemd manifest changes a lot between versions - build it in
code for sanity.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1436)

### CODE REVIEW

File: nodeup/pkg/model/docker.go [hunk 1]
```diff
@@ -268,5 +270,103 @@ func (b *DockerBuilder) Build(c *fi.ModelBuilderContext) error {
...
+
+	if oldDocker {
+		// Only in older versions of docker (< 1.12)
+		manifest.Set("Service", "MountFlags", "slave")
+	}
+
+	// Having non-zero Limit*s causes performance problems due to accounting overhead
+	// in the kernel. We recommend using cgroups to do container-local accounting.
+	// TODO: Should we set this?
+	//service.Set("Service", "LimitNOFILE", "infinity")
```
Comment (chrislovecnm) [inline nodeup/pkg/model/docker.go]: Not sure what the TODO means :(  Todo we need to set these to infinity?

File: nodeup/pkg/model/distribution.go [hunk 2]
```diff
@@ -31,17 +31,67 @@ var (
...
+	if d.IsSystemd() {
+		t = append(t, tags.TagSystemd)
+	}
+
+	return t
+}
+
+func (d Distribution) IsDebianFamily() bool {
+	switch d {
+	case DistributionJessie, DistributionXenial:
```
Comment (chrislovecnm) [inline nodeup/pkg/model/distribution.go]: Do we support any others?

File: nodeup/pkg/model/docker.go [hunk 3]
```diff
@@ -268,5 +270,103 @@ func (b *DockerBuilder) Build(c *fi.ModelBuilderContext) error {
...
+	manifest.Set("Install", "WantedBy", "multi-user.target")
+
+	manifestString := manifest.Render()
+
+	service := &nodetasks.Service{
+		Name:       "docker",
+		Definition: s(manifestString),
+	}
+
+	service.InitDefaults()
```
Comment (chrislovecnm) [inline nodeup/pkg/model/docker.go]: Should we add any logging in this?

File: nodeup/pkg/model/servicemanifest.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+func (s *ServiceManifestSection) Render() string {
+	var b bytes.Buffer
+
+	b.WriteString("[" + s.Key + "]\n")
+	for _, entry := range s.Entries {
+		b.WriteString(entry.Key + "=" + entry.Value + "\n")
+	}
+
+	return b.String()
+}
```
Comment (chrislovecnm) [inline nodeup/pkg/model/servicemanifest.go]: Any way to test any of this, or put in a issue for it?

File: nodeup/pkg/model/docker.go (hunk 1 above)
Comment (justinsb) [inline nodeup/pkg/model/docker.go]: I opened https://github.com/kubernetes/kubernetes/issues/39682 to figure out if we should set these to infinity

File: nodeup/pkg/model/distribution.go (hunk 2 above)
Comment (justinsb) [inline nodeup/pkg/model/distribution.go]: We have a default block that will warn us if/when more are added if we forget to add them here

File: nodeup/pkg/model/docker.go (hunk 3 above)
Comment (justinsb) [inline nodeup/pkg/model/docker.go]: What do you propose?

File: nodeup/pkg/model/servicemanifest.go (hunk 4 above)
Comment (justinsb) [inline nodeup/pkg/model/servicemanifest.go]: Sure ... please open an issue to do something comparable to the terraform testing we are doing.  We'd need to build an output representation for the tasks and check it.

File: nodeup/pkg/model/docker.go (hunk 1 above)
Comment (justinsb) [inline nodeup/pkg/model/docker.go]: Added link to the comment

File: nodeup/pkg/model/docker.go (hunk 3 above)
Comment (justinsb) [inline nodeup/pkg/model/docker.go]: Added logging of the manifest itself at v(8)

**Comment by chrislovecnm** [comment]
Waiting on e2e

**Pull request merged**